src/environment.jl

```diff
@@ -49,7 +49,15 @@ For more details, see [Finalizers](@ref).
 function refcount_dec()
     # refcount zero, all objects finalized, now finalize MPI
     if Threads.atomic_sub!(REFCOUNT, 1) == 1
-        !Finalized() && _Finalize()
+        if !Finalized()
+            # MPI can now be finalized, but MPI_Finalize is a collective and can act
+            # like a barrier (this may be implementation specific), if we are terminating
+            # due to a Julia exception, we should calling MPI_Finalize. We thus peek at the
+            # current exception, and only if that field is nothing do we terminate.
+            if ccall(:jl_current_exception, Any, ()) === nothing
```

Comment:
```suggestion
            if isempty(catch_backtrace())
```
would save using a `ccall`?

Reply:
Not sure, I am explicitly using the fact that the exception handler exited Julia instead of cleaning up and I am not sure in which state the backtrace stack is at that moment. So I think this is slightly more robust until we have proper support for this in Julia